subprojects/model-core/src/testFixtures/groovy/org/gradle/internal/reflect/validation/ValidationMessageChecker.groovy

```diff
@@ -31,35 +31,175 @@ trait ValidationMessageChecker {
+        String annotations = config.inConflict.collect { "@$it" }.join(", ")
+        config.render "has conflicting $config.kind: $annotations. The different annotations have different semantics and Gradle cannot determine which one to pick. Possible solution: Choose between one of the conflicting annotations."
+    }
+
+    String annotationInvalidInContext(@DelegatesTo(value = AnnotationContext, strategy = Closure.DELEGATE_FIRST) Closure<?> spec = {}) {
+        def config = display(AnnotationContext, 'annotation_invalid_in_context', spec)
+        config.render "is annotated with invalid property type @${config.annotation}. The '@${config.annotation}' annotation cannot be used in this context. Possible solution: Remove the property."
+    }
+
+    String missingAnnotationMessage(@DelegatesTo(value=MissingAnnotation, strategy=Closure.DELEGATE_FIRST) Closure<?> spec = {}) {
+        def config = display(MissingAnnotation, 'missing_annotation', spec)
+        config.render "is missing ${config.kind}. A property without annotation isn't considered during up-to-date checking. Possible solutions: Add ${config.kind} or mark it as @Internal."
+    }
+
+    String incompatibleAnnotations(@DelegatesTo(value=IncompatibleAnnotations, strategy=Closure.DELEGATE_FIRST) Closure<?> spec = {}) {
+        def config = display(IncompatibleAnnotations, 'incompatible_annotations', spec)
+        config.render "is annotated with @${config.annotatedWith} but that is not allowed for '${config.incompatibleWith}' properties. This modifier is used in conjunction with a property of type '${config.incompatibleWith}' but this doesn't have semantics. Possible solution: Remove the '@${config.annotatedWith}' annotation."
+    }
+
+    private <T extends ValidationMessageDisplayConfiguration> T display(Class<T> clazz, String docSection, @DelegatesTo(value = ValidationMessageDisplayConfiguration, strategy = Closure.DELEGATE_FIRST) Closure<?> spec = {}) {
+        def conf = clazz.newInstance(this)
+        conf.section = docSection
+        spec.delegate = conf
+        spec()
+        return (T) conf
+    }
+
+    static class IncompatibleAnnotations extends ValidationMessageDisplayConfiguration<IncompatibleAnnotations> {
+        String annotatedWith
+        String incompatibleWith
+
+        IncompatibleAnnotations(ValidationMessageChecker checker) {
+            super(checker)
+        }
+
+        IncompatibleAnnotations annotatedWith(String name) {
+            annotatedWith = name
+            this
+        }
+
+        IncompatibleAnnotations incompatibleWith(String name) {
+            incompatibleWith = name
+            this
+        }
+    }
+
+    static class MissingAnnotation extends ValidationMessageDisplayConfiguration<MissingAnnotation> {
+        String kind
+
+        MissingAnnotation(ValidationMessageChecker checker) {
+            super(checker)
+        }
+
+        MissingAnnotation kind(String kind) {
```

Comment:
I think it would be better to have extra methods for the kinds of annotations which are missing, instead of repeating `an input or output annotation` every time.

Reply:
Ok going to change that in the next PR since I got conflicts for this class.